### ISSUE TITLE
dbus dependency update

### ISSUE THREAD

**Issue by henry-nicolas** [hwchen/secret-service-rs#22]
Hello,

Is there any plans to upgrade the dbus dependency to something more recent (current version 0.2.3 was released back in 2016)?
That would help with downstream distribution in, ex., Linux distribution as the packaged dbus crate might not be that old.

Thanks in advance,

**Comment by hwchen** [comment]
I've been wondering if https://lib.rs/crates/zbus is mature enough. If it is, I might like to move to it since it's pure rust.

**Comment by henry-nicolas** [comment]
Great thanks for the quick feedback :)
It certainly displays as stable from the link you shared.

Maybe check with its author, @zeenix could share his thoughts?

**Comment by zeenix** [comment]
@henry-nicolas 👋 I guess it depends on what is mean by "stable enough". We're already API stable and don't break API unless it's absolutely unavoidable. We already have protects using our crates, and we've quite decent test coverage. 
I know 3 people who ported from dbus-rs to zbus and they were super happy with the results. I hope that helps.

If you've any further questions, I'd be more than happy to answer.

I've a favour to ask as well: if you do port your project, if you could document the process, that would be very helpful to others who would want to do the same.

Happy hacking! 😊

**Comment by emmanueltouzery** [comment]
For what it's worth, I needed just a little part of secret-service-rs in pure rust, and here is a working zbus solution for that tiny bit:

```rust
use serde::{Deserialize, Serialize};
use std::error::Error;
use zvariant_derive::Type;

use zbus::dbus_proxy;
use zvariant::{Str, Value};

// busctl --user --xml-interface introspect org.freedesktop.secrets /org/freedesktop/secrets > ~/secrets.xml
// cargo run --bin zbus-xmlgen ~/secrets.xml
#[dbus_proxy(
    interface = "org.freedesktop.Secret.Service",
    default_service = "org.freedesktop.secrets",
    default_path = "/org/freedesktop/secrets"
)]
trait Service {
    fn get_secrets(
        &self,
        items: &[zvariant::ObjectPath],
        session: &zvariant::ObjectPath,
    ) -> zbus::Result<SecretsResponse>;

    fn open_session(
        &self,
        algorithm: &str,
        input: &zvariant::Value,
    ) -> zbus::Result<(zvariant::OwnedValue, zvariant::OwnedObjectPath)>;

    fn search_items(
        &self,
        attributes: std::collections::HashMap<&str, &str>,
    ) -> zbus::Result<(
        Vec<zvariant::OwnedObjectPath>,
        Vec<zvariant::OwnedObjectPath>,
    )>;
}

#[derive(Debug, Serialize, Deserialize, Type)]
pub struct SecretsResponse(
    Vec<(
        zvariant::OwnedObjectPath,
        (zvariant::OwnedObjectPath, Vec<u8>, Vec<u8>, String),
    )>,
);

pub fn get_keyring_pass() -> Result<Option<String>, Box<dyn Error>> {
    let connection = zbus::Connection::new_session()?;

    let proxy = ServiceProxy::new(&connection)?;
    let (_val, session_path) = proxy.open_session("plain", &Value::Str(Str::from("")))?;

    let (unlocked, _locked) =
        proxy.search_items([("service", "projectpad-cli")].iter().cloned().collect())?;

    let p: &zvariant::ObjectPath = &unlocked[0];
    let SecretsResponse(secrets) = proxy.get_secrets(&[p.clone()][..], &session_path)?;
    Ok(secrets.first()
       .and_then(|s| std::str::from_utf8(&s.1.2).ok())
       .map(|s| s.to_string()))
}
```

Unfortunately this hits https://gitlab.freedesktop.org/zeenix/zbus/-/issues/104 so I needed [a tiny workaround in zbus for now](https://github.com/emmanueltouzery/zbus/commit/fb1825f7bea6f05a107e36a4aab80e69ff69d517).

**Comment by zeenix** [comment]
> Unfortunately this hits https://gitlab.freedesktop.org/zeenix/zbus/-/issues/104 so I needed [a tiny workaround in zbus for now](https://github.com/emmanueltouzery/zbus/commit/fb1825f7bea6f05a107e36a4aab80e69ff69d517).

Just FYI, this bug has been resolved and I'll make a release soon (< 3 days) with this fix included.

**Comment by zeenix** [comment]
> ```rust
> #[derive(Debug, Serialize, Deserialize, Type)]
> pub struct SecretsResponse(
>     Vec<(
>         zvariant::OwnedObjectPath,
>         (zvariant::OwnedObjectPath, Vec<u8>, Vec<u8>, String),
>     )>,
> );
> ```

As pointed out in the zbus issue, the return type of this method needs to be a `HashMap<OwnedObjectPath,(OwnedObjectPath, Vec<u8>, Vec<u8>, String)>`. Out of coincidence `OwnedObjectPath` isn't usable as a key with the last release and fix for this will also be part of the next release.

**Comment by henry-nicolas** [comment]
Thanks for the feedback @zeenix. Hopefuly this would help with @hwchen considering a switch to zbus crate, removing the dependency on the old dbus crate.

**Comment by bilelmoussaoui** [comment]
The migration to zbus could help with issues such #15 once zbus has async support. As a lot of GNOME apps rely on this crate, I can send a PR if you're okay with that

**Comment by hwchen** [comment]
@bilelmoussaoui sure, I'd be open to a PR. I'm also thinking about doing this work sometime around the end of the year, so please ping if you start so we don't duplicate effort.

**Comment by hwchen** [comment]
@bilelmoussaoui I've started work on a port to zbus. I think I should be able to handle it, but comments are welcome on the PR.

**Comment by hwchen** [comment]
Closed by #25 ! Version 2.0.0 (zbus plus some other breaking changes) is published.
